### DISCUSSION TITLE
"You are using pip version 9.0.3, ..." - "No I'm not ??"

### DISCUSSION THREAD

**Discussion by TigerEnteNBG** [netbox-community/netbox#5792] [Q&A] [answered]
Hi all,

there is something curious going on during the upgrades, as I always get a warning about my outdated pip version:

> 
> [Netbox]:/opt/netbox # pip install --upgrade pip
> Requirement already satisfied: pip in /usr/local/lib/python3.6/site-packages (21.0.1)
> 
> [Netbox]:/opt/netbox # ./upgrade.sh
> Removing old virtual environment...
> Creating a new virtual environment at /opt/netbox/venv...
> Installing Python system packages (pip3 install wheel)...
> Collecting wheel
>   Using cached https://files.pythonhosted.org/packages/65/63/39d04c74222770ed1589c0eaba06c05891801219272420b40311cd60c880/wheel-0.36.2-py2.py3-none-any.whl
> Installing collected packages: wheel
> Successfully installed wheel-0.36.2
> You are using pip version 9.0.3, however version 21.0.1 is available.
> You should consider upgrading via the 'pip install --upgrade pip' command.
> 

 
 Normally I don't bother as this came since we are using Netbox, but today I ran into this, and I really don't know how to solve it:
 

> Collecting cryptography>=2.5 (from paramiko>=2.6.0->napalm->-r local_requirements.txt (line 2))
>   Using cached https://files.pythonhosted.org/packages/27/5a/007acee0243186123a55423d49cbb5c15cb02d76dd1b6a27659a894b13a2/cryptography-3.4.4.tar.gz
>     Complete output from command python setup.py egg_info:
> 
>             =============================DEBUG ASSISTANCE==========================
>             If you are seeing an error here please try the following to
>             successfully install cryptography:
> 
>             Upgrade to the latest pip and try again. This will fix errors for most
>             users. See: https://pip.pypa.io/en/stable/installing/#upgrading-pip
>             =============================DEBUG ASSISTANCE==========================
> 
>     Traceback (most recent call last):
>       File "<string>", line 1, in <module>
>       File "/tmp/pip-build-6_mtj2cx/cryptography/setup.py", line 14, in <module>
>         from setuptools_rust import RustExtension
>     ModuleNotFoundError: No module named 'setuptools_rust'
> 
>     ----------------------------------------
> Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-6_mtj2cx/cryptography/
> You are using pip version 9.0.3, however version 21.0.1 is available.
> You should consider upgrading via the 'pip install --upgrade pip' command.
> 

Any Ideas ?

We are running on CentOS8 and Netbox v2.10.3.

**Comment by candlerb** [comment, marked as answer]
I have this in my `local_requirements.txt`:

```
setuptools>=36
napalm
napalm-ros
```

Since it looks like you're also using napalm, see if the `setuptools>=36` line fixes it for you.

**Reply by TigerEnteNBG** [reply]
You are right. `setuptools>=36` and `napalm-ros` were missing in your local_requierments. Thanks for the hint

But unfortunately, that does not fix my issue.

**Reply by candlerb** [reply]
`napalm-ros` is only for Mikrotik devices.

I haven't seen this `setuptools_rust` error before. Can you show a way to reproduce your problem: in particular, exactly what you you have in your `local_requirements.txt`?  Is there anything else that you customized?

**Reply by candlerb** [reply]
I am able to reproduce your problem by creating a CentOS 8 lxd container under Ubuntu 18.04, and following the [standard install instructions](https://netbox.readthedocs.io/en/stable/installation/3-netbox/) with napalm:

```
$ lxc launch images:centos/8/amd64 nbtest
$ lxc exec nbtest bash
[root@nbtest ~]# yum install -y gcc python36 python36-devel python3-pip libxml2-devel libxslt-devel libffi-devel openssl-devel redhat-rpm-config
...
[root@nbtest ~]# mkdir -p /opt/netbox/ && cd /opt/netbox/
[root@nbtest netbox]# yum install -y git
...
[root@nbtest netbox]# git clone -b master https://github.com/netbox-community/netbox.git .
...
[root@nbtest netbox]# echo "napalm" >local_requirements.txt
[root@nbtest netbox]# /opt/netbox/upgrade.sh
... lots of packages installed
Collecting cryptography>=2.5 (from paramiko>=2.6.0->napalm->-r local_requirements.txt (line 1))
  Downloading https://files.pythonhosted.org/packages/27/5a/007acee0243186123a55423d49cbb5c15cb02d76dd1b6a27659a894b13a2/cryptography-3.4.4.tar.gz (545kB)
    100% |████████████████████████████████| 552kB 2.3MB/s
    Complete output from command python setup.py egg_info:

            =============================DEBUG ASSISTANCE==========================
            If you are seeing an error here please try the following to
            successfully install cryptography:

            Upgrade to the latest pip and try again. This will fix errors for most
            users. See: https://pip.pypa.io/en/stable/installing/#upgrading-pip
            =============================DEBUG ASSISTANCE==========================

    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-iejeken3/cryptography/setup.py", line 14, in <module>
        from setuptools_rust import RustExtension
    ModuleNotFoundError: No module named 'setuptools_rust'

    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-iejeken3/cryptography/
You are using pip version 9.0.3, however version 21.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

It looks like the problem is that as the error says, pip is too old - but local_requirements.txt is installed *after* requirements.txt which is too late to fix it there.

The following diff fixes it for me:

```
--- upgrade.sh.orig	2021-02-11 18:34:29.091900888 +0000
+++ upgrade.sh	2021-02-11 18:46:11.481406965 +0000
@@ -30,7 +30,7 @@
 source "${VIRTUALENV}/bin/activate"

 # Install necessary system packages
-COMMAND="pip3 install wheel"
+COMMAND="pip3 install wheel 'pip>=21'"
 echo "Installing Python system packages ($COMMAND)..."
 eval $COMMAND || exit 1

```

However, it would be up to @jeremystretch to decide if this is the correct final solution or not.  At least he should be able to reproduce the issue easily now.

**Reply by TigerEnteNBG** [reply]
That works like a charm, thanks for digging into this :)

**Reply by jeremystretch** [reply]
The solution, as the error message suggests, is to simply upgrade pip. The virtual environment is created using the system's version of pip, which in this case is too old. We've worked around this in #5776 by adding a step in `upgrade.sh` to upgrade pip within the virtual environment automatically.

**Reply by candlerb** [reply]
Based on my testing, even with #5776 changes to `./upgrade.sh`, under CentOS 8 it's also necessary to change

```
COMMAND="pip install --upgrade pip"
```

to

```
COMMAND="pip install --upgrade pip setuptools wheel"
```

**Comment by rchaud** [comment]
This is an issue with pip version. Cryptography requires pip version 19.0 and higher. Upgrading your pip version should work. I notice that python3-pip comes installed with version 9.0.3 by default in CentOS.

**Reply by TigerEnteNBG** [reply]
Hi chniochao,

as I wrote in my original post an topic:

[Netbox]:/opt/netbox # pip install --upgrade pip
Requirement already satisfied: pip in /usr/local/lib/python3.6/site-packages (21.0.1)

I had the lastest version installed but it was not recogonized by the setup.

So you and some other told me to "just upgrade pip", please advice how, if that command doesn't do the trick.

candlerB's solution worked for me as a charme.

**Reply by candlerb** [reply]
It's the version *inside* the virtualenv that matters, not the one outside.  Hence the `pip install --upgrade pip` has to be done while the virtualenv is activated.

**Reply by linehammer** [reply]
[Upgrade pip ](http://net-informations.com/python/iq/pip.htm)

`python -m pip install --upgrade pip`

for Python 3:

`python3 -m pip install --upgrade pip`

for Python 2:

`python2 -m pip install --upgrade pip`

**Reply by candlerb** [reply]
This shouldn't be necessary any more, `./upgrade.sh` handles it for you.
